src/main/java/frc/robot/subsystems/Climber.java

```diff
@@ -0,0 +1,117 @@
+    }
+
+    private void setLiftForces(double chainRoll) {
+        //TODO: add pid calculation and ff to each side lift force
+    }
+
+    private double getChainHeight(Translation2d robotTranslation, double armLoc) {
+        Chain chain;
+
+        if (isBlueAlliance()) {
+            if (Chain.CHAIN_0.inRange(robotTranslation)) chain = Chain.CHAIN_0;
+            else if (Chain.CHAIN_120.inRange(robotTranslation)) chain = Chain.CHAIN_120;
+            else chain = Chain.CHAIN_240;
+        } else {
+            if (Chain.CHAIN_60.inRange(robotTranslation)) chain = Chain.CHAIN_60;
+            else if (Chain.CHAIN_180.inRange(robotTranslation)) chain = Chain.CHAIN_180;
+            else chain = Chain.CHAIN_300;
+        }
+
+        Translation2d projection = Chain.getProjection(robotTranslation, chain);
+        double armProjection = CHAIN_LENGTH_IN_XY_METERS / 2 - chain.getPosEdge().getDistance(projection) + armLoc;
+        return CHAIN_PARABOLA_PARAMETER * Math.pow(armProjection,2) + MINIMAL_CHAIN_HEIGHT_METERS;
+    }
+
+
+
+    private double getDesiredHeight() {
+        return Math.min(leftSide.getHeight(), rightSide.getHeight()) - MINIMAL_HEIGHT;
+    }
+
+    private class ClimberSide {
+        private Neo motor;
+        private TrapezoidProfile trapezoidProfile;
+        private Timer timer;
+        private double liftingForce = 0;
+
+        private ClimberSide(Gains motorGains, int motorID) {
+            motor = new Neo(motorID, motorGains);
+            motor.setConversionFactors(kRotToMeter, kRotToMeterPerSec);
+        }
+
+        private Command setLengthCommand(double length) {
+
+            return new FunctionalCommand(
+                    () -> {
+                        timer.restart();
+                        trapezoidProfile = new TrapezoidProfile(new TrapezoidProfile.Constraints(kMaxLinearVelocity, kMaxLinearAcceleration));
+                    },
+                    () -> {
+                        TrapezoidProfile.State currentState = new TrapezoidProfile.State(motor.getPosition(), motor.getVelocity());
+                        TrapezoidProfile.State goalState = new TrapezoidProfile.State(length, 0);
+
+                        TrapezoidProfile.State setpoint = trapezoidProfile.calculate(timer.get(), currentState, goalState);
+                        double feedForward = Math.signum(setpoint.velocity) * motor.getGains().ks +
+                                motor.getGains().kv * currentState.velocity + motor.getGains().kg + liftingForce;
+
+                        motor.setReference(setpoint.position, CANSparkBase.ControlType.kPosition, 0, feedForward);
+                    },
+                    (__) -> timer.stop(),
+                    () -> timer.hasElapsed(trapezoidProfile.totalTime())
+            );
+        }
+
+        private void setLiftingForce(double liftingForce) {
+            this.liftingForce = liftingForce;
+        }
```

Comment:
What is the purpose of this (the setter as well as the field itself) ?
Documentation.

Reply:
The purpose is to adjust to the forces applied to the climber when climbing, and the robot's weight is carried , I already have the physics to calculate it using the gyro, I only need to update it